matching/README.md

```diff
@@ -0,0 +1,991 @@
+  replace fringe with specialized fringe
+```
+
+### Default function
+`Default` is a function that takes a column j of a matrix and a matrix m and
+returns a decomposed matrix representing the matrix after having determined
+that the term at position o_j does not have any of the constructors in
+signature(s_j, column j of m).
+
+`Default` operates by taking each row i of the matrix and either deleting it,
+or transforming it into a new row i', depending on the structure of the pattern
+P_ij.
+
+We define default first on rows.
+
+* default(i, X) = [P_i1, ..., P_i(j-1), P_i(j+1), ..., P_im] -> C
+  where C is the clause of row i except S_i is updated with the binding
+  X |-> (o_j, s_j)
+* default(i, c(p_1, ..., p_a)) = None
+* default(i, p as X) = default(i, p) except S_i is updated with the binding
+  X |-> (o_j, s_j)
+
+Default operations on maps, lists, and sets are described in their respective
+sections.
+
+We then also must define specialization on occurrence and sort vectors:
+
+* default(i, [o_1, ..., o_m]) = [o_1, ..., o_(j-1), o_(j+1), ..., o_m]
+* default(i, [s_1, ..., s_m]) = [s_1, ..., s_(j-1), s_(j+1), ..., s_m]
+
+We then define default on matrices using the following pseudocode:
+
+```
+def default(Matrix m, Column j): Matrix
+  if signature(s_j, column j of m) contains every constructor of sort s_j:
+    return None
+  if s_j is a normal sort, a map sort, or a set sort:
+    for i = 1 to n:
+      val newRow = default(i, P_ij)
+      if newRow is None:
+        delete row i
+      else:
+        replace row i with newRow
+  else: //s_j is a list sort
+    val (maxHead, maxTail) = maxListSize(column j)
+    return specialize(m, j, Len(maxHead + maxTail))
+```
+
+`maxListSize` is defined in the section on list matching.
+
+### Semantics of decision trees
+
+A decision tree alone is not meaningful; it has meaning only insofar as it is
+executable at runtime. We define the function `Evaluate` which takes a decision
+tree and a concrete pattern (i.e., a pattern with no variables), and either
+returns an action A and a substitution S mapping from occurrences to
+concrete patterns, or the special value Fail.
+
+Decision trees are not actually implemented this way in our implementation.
+It is vastly more efficient to transform a decision tree into a native function
+that takes a term as input in some internal representation and either raises
+an exception (on failure) or calls a function which returns the right-hand side
+of rule A substituted with S. However, we define it this way formally in order
+to allow implementations to choose the details of exactly what machine code
+to generate.
+
+The function `evaluate` is defined statefully using an intermediate state
+representing the substitution S which initially maps [] to the initial pattern,
+and a stack of (DT, S) pairs that is initially empty. In order to
+implement set/map matching, S is extended to allow us to map occurrences to an
+iterator over a map or a set, or to None.
```

Comment:
What does it mean for us to map an occurrence to an iterator? I can visualise a simple substitution but it's less clear to me what the semantics of this are. Why don't lists participate in iterator mapping - is it to do with the ACU property?

Reply:
We don't need to use iterators to implement list matching. Essentially, you can think of occurrences as being the variables that exist at runtime. Mapping an occurrence to an iterator just means that instead of storing a term, we store an iterator.